core/query/query.ml

```diff
@@ -715,13 +648,94 @@ struct
+          norm_comp env body
+    | Q.Primitive "AsList", [xs] ->
+        xs
+    | Q.Primitive "Cons", [x; xs] ->
+        reduce_concat env [Q.Singleton x; xs]
+    | Q.Primitive "Concat", ([_xs; _ys] as l) ->
+        reduce_concat env l
+    | Q.Primitive "ConcatMap", [f; xs] ->
+        begin
+          match f with
+            | Q.Closure (([x], body), closure_env) ->
+                let env = env ++ closure_env in
+                  reduce_for_source env
+                    (xs, fun v -> let env = bind env (x, v) in norm_comp env body)
+            | _ -> assert false
+        end
+    | Q.Primitive "Map", [f; xs] ->
+        begin
+          match f with
+            | Q.Closure (([x], body), closure_env) ->
+                let env = env ++ closure_env in
+                  reduce_for_source env
+                    (xs, fun v -> let env = bind env (x, v) in Q.Singleton (norm_comp env body))
+            | _ -> assert false
+        end
+    | Q.Primitive "SortBy", [f; xs] ->
+        begin
+          match xs with
+            | Q.Concat [] -> Q.Concat []
+            | _ ->
+                let gs, os', body =
+                  match xs with
+                    | Q.For (_, gs, os', body) -> gs, os', body
+                    | Q.Concat (_::_)
+                    | Q.Singleton _
+                    | Q.Table _ ->
+                        (* I think we can omit the `Table case as it
+                           can never occur *)
+                        (* eta-expand *)
+                        eta_expand_list xs
+                    | _ -> assert false in
+                let xs = Q.For (None, gs, os', body) in
+                  begin
+                    match f with
+                      | Q.Closure (([x], os), closure_env) ->
+                          let os =
+                            let env = bind (env ++ closure_env) (x, Q.tail_of_t xs) in
+                              let o = norm_comp env os in
+                                match o with
+                                  | Q.Record fields ->
+                                      List.rev (StringMap.fold (fun _ o os -> o::os) fields [])
+                                  | _ -> assert false
+                          in
+                            Q.For (None, gs, os @ os', body)
+                      | _ -> assert false
+                  end
+        end
+    | Q.Primitive "not", [v] ->
+      reduce_not env (v)
+    | Q.Primitive "&&", [v; w] ->
+      reduce_and env (v, w)
+    | Q.Primitive "||", [v; w] ->
+      reduce_or env (v, w)
+    | Q.Primitive "==", [v; w] ->
+      reduce_eq env (v, w)
+    | Q.Primitive f, args ->
+        Q.Apply (Q.Primitive f, args)
+    | Q.If (c, t, e), args ->
+        reduce_if_condition env (c, apply env (t, args), apply env (e, args))
+    | Q.Apply (f, args), args' ->
+        apply env (f, args @ args')
+    | t, _ -> eval_error "Application of non-function: %s" (string_of_t t)
+
+  and reduce_concat env vs =
```

Comment:
The various reduce_ fucntions were previously not really mutually recursive with translation/normalization.  Now they are mutually recursive with `norm,` but only for the trivial reason that they first call `norm` on their arguments, which could easily be avoided by moving these calls back into `norm` itself where the `reduce_` functions are called.  There is also no other particular reason `env` needs to be threaded through the `reduce_` functions other than that `norm` is being called and requires `env`.  The `reduce_` functions could be moved earlier, defined before / separatelt from the mutual recursion with `norm` and `env` removed as an argument.   Indeed I think they could even live in `Query.Lang`.  I think that would be cleaner, even though there is then a bit more clutter when `reduce_` functions are called.

Reply:
This seems plausible, I'll try it.